pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java

```diff
@@ -0,0 +1,590 @@
+                case "m" -> TimeUnit.MILLISECONDS;
+                case "u" -> TimeUnit.MICROSECONDS;
+                case "n" -> TimeUnit.NANOSECONDS;
+                // This should NEVER be reachable, because the matcher would not have matched.
+                default -> throw new GrpcException(GrpcStatus.INTERNAL, "Invalid unit: " + unit);
+            });
+            return deadlineDetector.scheduleDeadline(deadline, () -> {
+                route.deadlineExceededCounter().increment();
+                incoming.onError(new GrpcException(GrpcStatus.DEADLINE_EXCEEDED));
+            });
+        }
+
+        return new NoopScheduledFuture();
+    }
+
+    /**
+     * According to the specification, the "Response-Headers" are:
+     *
+     * <pre>HTTP-Status [Message-Encoding] [Message-Accept-Encoding] Content-Type *Custom-Metadata</pre>
+     *
+     * <p>Where "Status" is <strong>always</strong> 200 OK.</p>
+     *
+     * <p>The Response-Headers are normally sent, followed by the data, followed by trailers. But if the request fails
+     * right away, before any handling, then it is possible to send Trailers-Only instead.
+     */
+    private void sendResponseHeaders(
+            @Nullable final Header messageEncoding,
+            @NonNull final HttpMediaType contentType,
+            @NonNull final List<Header> customMetadata) {
+
+        // Some headers are http2 specific, the rest are used for the grpc protocol
+        final var grpcHeaders = WritableHeaders.create();
+        final var http2Headers = Http2Headers.create(grpcHeaders);
+        // FUTURE: I think to support custom headers in the response, we would have to list them here.
+        // Since this has to be sent before we have any data to send, we must know ahead of time which
+        // custom headers are to be returned.
+        grpcHeaders.set(HeaderNames.TRAILER, "grpc-status, grpc-message");
+        http2Headers.status(Status.OK_200);
+        grpcHeaders.contentType(contentType);
+        grpcHeaders.set(GRPC_ACCEPT_ENCODING, IDENTITY);
+        customMetadata.forEach(grpcHeaders::set);
+        if (messageEncoding != null) {
+            grpcHeaders.set(messageEncoding);
+        }
+
+        streamWriter.writeHeaders(http2Headers,
+                streamId,
+                Http2Flag.HeaderFlags.create(Http2Flag.END_OF_HEADERS),
+                flowControl.outbound());
+    }
+
+    /**
+     * A convenience class for building the trailers. In the specification, it says:
+     *
+     * <pre>
+     *     Trailers → Status [Status-Message] *Custom-Metadata
+     *     Status → "grpc-status" 1*DIGIT ; 0-9
+     *     Status-Message → "grpc-message" Percent-Encoded
+     *     Percent-Encoded → 1*(Percent-Byte-Unencoded / Percent-Byte-Encoded)
+     *     Percent-Byte-Unencoded → 1*( %x20-%x24 / %x26-%x7E ) ; space and VCHAR, except %
+     *     Percent-Byte-Encoded → "%" 2HEXDIGIT ; 0-9 A-F
+     * </pre>
+     */
+    private class TrailerBuilder {
```

Comment:
Ah, the magic of nested classes... I see...
Given that the usage above looks a bit confusing, I'd prefer if this was an outer (or static inner) class. But I'm okay if you really want the code to look like magic.

Reply:
See my comment [above](https://github.com/hashgraph/pbj/pull/259#discussion_r1765501729), I'm leaning towards leaving the magic for now.